velox/functions/sparksql/String.h

```diff
@@ -1431,4 +1432,208 @@ struct LevenshteinDistanceFunction {
+      const arg_type<Varchar>* inputPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        StringView{kMaskedUpperCase_},
+        StringView{kMaskedLowerCase_},
+        StringView{kMaskedDigit_},
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        StringView{kMaskedLowerCase_},
+        StringView{kMaskedDigit_},
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        StringView{kMaskedDigit_},
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr,
+      const arg_type<Varchar>* digitCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        getMaskedChar(digitCharPtr),
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr,
+      const arg_type<Varchar>* digitCharPtr,
+      const arg_type<Varchar>* otherCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        getMaskedChar(digitCharPtr),
+        getMaskedChar(otherCharPtr));
+    return true;
+  }
+
+ private:
+  void doCall(
+      out_type<Varchar>& result,
+      StringView input,
+      const std::optional<StringView> upperChar,
+      const std::optional<StringView> lowerChar,
+      const std::optional<StringView> digitChar,
+      const std::optional<StringView> otherChar) const {
+    auto inputBuffer = input.data();
+    const size_t inputSize = input.size();
+    result.reserve(inputSize);
+    auto outputBuffer = result.data();
+    size_t inputIdx = 0;
+    size_t outputIdx = 0;
+    while (inputIdx < inputSize) {
+      utf8proc_int32_t curCodePoint;
+      int charByteSize;
```

Comment:
How about removing `charByteSize` and use `maskedCharByteSize` instead as `charByteSize` is only used to determine `maskedCharByteSize`?

Reply:
not true. CharByteSize is use for initializing `maskedCharByteSize`  and also update `inputIdx` - https://github.com/gayangya/velox/blob/gayangya/spark_mask/velox/functions/sparksql/String.h#L1597

Reply:
Got it. Thanks.